### ISSUE TITLE
Loading a PNG with alpha and rendering not working correctly on Windows GL

### ISSUE THREAD

**Issue by battlebottle** [mono/MonoGame#1946]
This issue affect MonoGame on Windows GL but may affect other platforms too. It affects MonoGame 3.0, and also 3.2.

Basically loading a texture with this code (F#):

```
use file = System.IO.File.OpenRead("testTexture.png")
this.texture <- Texture2D.FromStream(this.GraphicsDevice, file)
```

and rendering with this code:

    this.spriteBatch.Begin(SpriteSortMode.Immediate, BlendState.NonPremultiplied);
    this.spriteBatch.Draw(this.texture, Vector2.Zero, Color.White)
    this.spriteBatch.End()

will reproduce the error. This code works fine and renders fine when using the XNA4 libraries, but has weird alpha blending artefacts when using MonoGame.

Here's some renderings to illustrate the issue. First the correct renderings using the XNA libraries
![Image](http://i.imgur.com/2Rtqhuk.png)

Now the bad renderings using MonoGame:

![Image](http://i.imgur.com/Dny26gI.png)

I've tried to figure out the nature of the issue, but havne;t gotten very far besides realising that it's not a simple case of `BlendState.NonPremultiplied` being ignored or something similar. For examples, on analysis I've noticed the green colour channel blends differently than the red and blue colour channels among other odd things. I've creates a large image that illustrates the issue I've noticed in what is hopefully a useful way:

http://i.imgur.com/CEUQqm0.png

Here's the project files for reproducing the bad rendering (F# project):
http://www.qfpost.com/file/d?g=EhZh52Zjc

I also made a post about this issue on stack overflow which may be worth monitoring:
http://stackoverflow.com/questions/18320633/loading-texture-form-png-in-monogame-transaprency-issue

**Comment by Draih** [comment]
This issue is still not fixed in Monogame 3.3. I'm attempting to render a transparent .PNG file which is primarily red, and in the running file, the image comes out purple. Checking the repository, the change to fix this issue never made it in.

**Comment by fjeandinot** [comment]
Hi ! This issue is still not fixed in Monogame 3.4.

As you can see, this is the result with WindowsGL (classic png + BlendState.NonPremultiplied)
(the center of the picture is a sample for comparaison)
![bad](https://cloud.githubusercontent.com/assets/17014348/12732742/35a6bfd6-c937-11e5-93ad-5be3e29d57b3.png)

And the result expected
![good](https://cloud.githubusercontent.com/assets/17014348/12732743/35a7a040-c937-11e5-9e4a-75ecfe1b3320.png)

The "grood" result can be obtained with WindowsDX + a premultiplied png and BlendState.AlphaBlend so I guess it is a GL specific issue...

**Comment by tomspilman** [comment]
@fjeandinot - Can you test with the latest development build of 3.5?  We though that https://github.com/mono/MonoGame/issues/2424 resolved this.

**Comment by fjeandinot** [comment]
Hi tom !

Still the same result. Sorry.

**Comment by tomspilman** [comment]
@fjeandinot - Are your textures being processed by the content pipeline or being loaded from PNGs?  Can you share a simple test project that exhibits the error?

**Comment by fjeandinot** [comment]
The picture is loaded from a png processed by inkscape.
I used to process the texture with the MonoGame Pipeline but the xnb files where too large and the dxt compression deteriorate my sprites.

Here's my tesing code :

`
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;

namespace View
{
	public class Viewer : Game
	{
		GraphicsDeviceManager graphics;
		SpriteBatch spriteBatch;
		Texture2D texture;
		Viewport viewport;

		public Viewer ()
		{
			graphics = new GraphicsDeviceManager (this);
			Content.RootDirectory = "Content";	            
		}

		protected override void Initialize ()
		{
			const int width = 1024;
			const int height = 1024;

			Window.Position = new Point ((GraphicsDevice.DisplayMode.Width / 2) - (width / 2),
				(GraphicsDevice.DisplayMode.Height / 2) - (height / 2));
			graphics.PreferredBackBufferWidth = width;
			graphics.PreferredBackBufferHeight = height;
			graphics.ApplyChanges ();

			viewport = new Viewport (0, 0, width, height);

			base.Initialize ();
		}

		protected override void LoadContent ()
		{
			texture = Content.Load<Texture2D> ("spritesheet.png");
			spriteBatch = new SpriteBatch (GraphicsDevice);
		}

		protected override void Update (GameTime gameTime)
		{
			if (texture == null) {
				this.Exit ();
			}
			base.Update (gameTime);
		}

		protected override void Draw (GameTime gameTime)
		{
			graphics.GraphicsDevice.Clear (Color.White);
			graphics.GraphicsDevice.Viewport = viewport;
			spriteBatch.Begin (SpriteSortMode.Deferred, BlendState.NonPremultiplied);
			spriteBatch.Draw (texture, Vector2.Zero, Color.White);
			spriteBatch.End ();
			base.Draw (gameTime);
		}
	}
}

`

Do you want a zipped version of the project with the png ?

**Comment by fjeandinot** [comment]
[https://www.dropbox.com/s/zpfkmluehareugs/View4.zip?dl=0](url)

**Comment by fjeandinot** [comment]
[`Link to my Viewer`](https://www.dropbox.com/s/zpfkmluehareugs/View4.zip?dl=0)

**Comment by fjeandinot** [comment]
If the developpement build is recreated under the next 24 hour, I'll be glad to test every flavor of monogame with the premultiplied and the non premultiplied png with each option. As the viewer produce a direct result, it could be a cool non-regression test.

**Comment by tomspilman** [comment]
> If the developpement build is recreated under the next 24 hour

The new build is already up...  we update the website on every merge to the develop branch.  So if you grab the current version it will have the fix in #4487.

> it could be a cool non-regression test.

Seems like we could make a simpler test.  If the issue is just one of `Texture2D.FromStream()` not loading textures correctly... we can test it by just examining the pixels of the loaded texture.  No rendering required.

I see we have no unit test for `Texture2D.FromStream()`... so this would be a good start to rectifying that.

**Comment by tomspilman** [comment]
@fjeandinot - FYI...   @Tzenchor is adding some unit tests for `Texture2D.FromStream()` in https://github.com/mono/MonoGame/pull/4493.  You might want to wait for his PR if you planned to work on your own.

**Comment by fjeandinot** [comment]
Hi ! I'have tested the DesktopGL ans the Windows

Here's the result for the non premultiplied alpha png :
- DesktopGL + BlendState.AlphaBlend => white screen
- DesktopGL + BlendState.NonPremultiplied => white screen
- Windows + BlendState.AlphaBlend => perfect !!!
- Windows + BlendState.NonPremultiplied => gray borders

Here's the result for the premultiplied alpha png :
- DesktopGL + BlendState.AlphaBlend => white screen
- DesktopGL + BlendState.NonPremultiplied => white screen
- Windows + BlendState.AlphaBlend => perfect !!!
- Windows + BlendState.NonPremultiplied => gray borders

Something's probably broken with the GL version, but isn't it strange that the windows version act that way ? 

I think I'm missing a point about the differents version of monogame...

**Comment by tomspilman** [comment]
> I think I'm missing a point about the differents version of monogame

DesktopGL uses OpenGL to render.  Windows uses DirectX.  So very different code and graphics layers to the same hardware.

**Comment by fjeandinot** [comment]
Hi ! Since you have fixed this problem, the GL version seems to display "nothing" from a png regadless of the BlendState or the png alpha mode.

**Comment by fjeandinot** [comment]
Hi,

I keep updating my version of monogame by downloading the daily build but the GL version keep not showing raw png...

**Comment by KonajuGames** [comment]
> I keep updating my version of monogame by downloading the daily build but the GL version keep not showing raw png.

I replicated the problem and created a fix in #4609.
